### PR TITLE
Add aten::upsample_bicubic2d and its variant

### PR DESCRIPTION
**Pull request by majing921201**
no case need to be added for fine grain test

### CODE REVIEW

File: src/aten/sycl/UpSample.h [hunk 1]
```diff
@@ -0,0 +1,130 @@
+#pragma once
+
+#include <ATen/TensorUtils.h>
+#include <math.h>
+
+namespace at::native::xpu {
+
+template <typename scalar_t>
+inline scalar_t min(scalar_t a, scalar_t b) {
```
Comment (fengyuan14) [inline src/aten/sycl/UpSample.h:9]: Can we you std func?

Comment (fengyuan14) [reply src/aten/sycl/UpSample.h:9]: Or we are aligning with CUDA?

**Comment by fengyuan14** [comment]
Are the extended cases enabled?

File: src/aten/sycl/UpSampleBicubic2dKernels.cpp [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+#include <ATen/AccumulateType.h>
+#include <ATen/Dispatch.h>
+#include <ATen/TensorUtils.h>
+#include <comm/SYCLContext.h>
+#include "UpSample.h"
+
+namespace at::native::xpu {
+
+template <typename scalar_t, typename accscalar_t>
+struct UpsampleBicubic2dOutFrameKernelFunctor {
```
Comment (fengyuan14) [inline src/aten/sycl/UpSampleBicubic2dKernels.cpp:11]: Remove out from the name of all kernels.

File: src/aten/sycl/UpSampleBicubic2dKernels.cpp [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+            odata[n][c][output_y][output_x] = val;
+          }
+        }
+        return;
+      }
+
+      // Interpolation kernel
+      accscalar_t real_x = area_pixel_compute_source_index(
+          width_scale_, output_x, align_corners_, /*cubic=*/true);
+      // TODO: floor should support at dispatch to half type
```
Comment (fengyuan14) [inline src/aten/sycl/UpSampleBicubic2dKernels.cpp:38]: Can you evaluate the issue?

File: src/aten/sycl/UpSampleBicubic2dKernels.cpp [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+                  coefficients[0],
+                  coefficients[1],
+                  coefficients[2],
+                  coefficients[3],
+                  t_y));
+        }
+      }
+    }
+  }
+  UpsampleBicubic2dOutFrameKernelFunctor(
```
Comment (fengyuan14) [inline src/aten/sycl/UpSampleBicubic2dKernels.cpp:97]: UpsampleBicubic2dKernelFunctor. It is our flavor.

File: src/aten/sycl/UpSampleBicubic2dKernels.cpp [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    PackedTensorAccessor64<scalar_t, 4> odata,
+    const PackedTensorAccessor64<const scalar_t, 4> idata,
+    int64_t onum,
+    bool align_corners,
+    const accscalar_t height_scale,
+    const accscalar_t width_scale) {
+  auto queue = getCurrentSYCLQueue();
+  int64_t local_range = static_cast<int64_t>(1);
+  int64_t global_range = static_cast<int64_t>(1);
+  if (onum != 0) {
```
Comment (fengyuan14) [inline src/aten/sycl/UpSampleBicubic2dKernels.cpp:131]: Why global range == 1, when onum == 0. This means we still launch kernel for onum==0?

File: src/aten/sycl/UpSample.h (hunk 1 above)
Comment (majing921201) [reply src/aten/sycl/UpSample.h:9]: > Or we are aligning with CUDA?

yes, aligned

File: src/aten/sycl/UpSampleBicubic2dKernels.cpp (hunk 3 above)
Comment (majing921201) [reply src/aten/sycl/UpSampleBicubic2dKernels.cpp:38]: > Can you evaluate the issue?

already used c++ standard floorf(), follow cuda, forget to remove this comment

File: src/ATen/native/xpu/sycl/UpSampleBicubic2dKernels.cpp [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+#include <ATen/Dispatch.h>
+#include <ATen/TensorUtils.h>
+#include <ATen/ceil_div.h>
+#include <comm/SYCLContext.h>
+#include "UpSample.h"
+
+namespace at::native::xpu {
+
+template <typename scalar_t, typename accscalar_t>
+struct UpsampleBicubic2dOutFrameKernelFunctor {
```
Comment (fengyuan14) [inline src/ATen/native/xpu/sycl/UpSampleBicubic2dKernels.cpp:12]: Remove out from the name

File: src/ATen/native/xpu/sycl/UpSampleBicubic2dKernels.cpp [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+            odata[n][c][output_y][output_x] = val;
+          }
+        }
+        return;
+      }
+
+      // Interpolation kernel
+      accscalar_t real_x = area_pixel_compute_source_index(
+          width_scale_, output_x, align_corners_, /*cubic=*/true);
+      // TODO: floor should support at dispatch to half type
```
Comment (fengyuan14) [inline src/ATen/native/xpu/sycl/UpSampleBicubic2dKernels.cpp:39]: What's the comment mean?

File: src/ATen/native/xpu/UpSampleBicubic2d.cpp [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+#include <ATen/xpu/XPUNativeFunctions.h>
+
+namespace at {
+Tensor& XPUNativeFunctions::upsample_bicubic2d_out(
+    const Tensor& self,
+    IntArrayRef output_size,
+    bool align_corners,
+    std::optional<double> scales_h,
+    std::optional<double> scales_w,
+    Tensor& output) {
```
Comment (fengyuan14) [inline src/ATen/native/xpu/UpSampleBicubic2d.cpp:13]: Missing meta process.
image

File: src/ATen/native/xpu/XPUFallback.template [hunk 9]
```diff
@@ -423,6 +422,7 @@ TORCH_LIBRARY_IMPL(aten, XPU, m) {
     "var_mean.correction",
     "vdot",
     "xlogy.OutTensor",
+    "_upsample_bicubic2d_aa.out",
```
Comment (fengyuan14) [inline src/ATen/native/xpu/XPUFallback.template:425]: For enabling test case?

File: src/ATen/native/xpu/sycl/UpSample.h [hunk 10]
```diff
@@ -9,6 +9,9 @@
 
 #include <math.h>
 
+#include <ATen/TensorUtils.h>
+#include <math.h>
```
Comment (fengyuan14) [inline src/ATen/native/xpu/sycl/UpSample.h:13]: Got it in line10?

File: src/ATen/native/xpu/sycl/UpSampleBicubic2dKernels.cpp [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+
+void upsample_bicubic2d_kernel(
+    Tensor& output,
+    const Tensor& input,
+    IntArrayRef output_size,
+    bool align_corners,
+    std::optional<double> scales_h,
+    std::optional<double> scales_w) {
+  TensorArg input_arg{input, "input", 1}, output_arg{output, "output", 2};
+  checkAllSameGPU(__func__, {input_arg, output_arg});
```
Comment (fengyuan14) [inline src/ATen/native/xpu/sycl/UpSampleBicubic2dKernels.cpp:146]: If they are meta process, please move them to op level. Since they are host only codes.

File: src/ATen/native/xpu/sycl/UpSampleBicubic2dKernels.cpp [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+
+  output.zero_();
+
+  const int num_output_elements = output_height * output_width;
+
+  AT_DISPATCH_FLOATING_TYPES_AND2(
+      at::ScalarType::Half,
+      at::ScalarType::BFloat16,
+      input.scalar_type(),
+      "upsample_bicubic2d_kernel",
```
Comment (fengyuan14) [inline src/ATen/native/xpu/sycl/UpSampleBicubic2dKernels.cpp:162]: `upsample_bicubic2d_xpu`

File: src/ATen/native/xpu/sycl/UpSampleBicubic2dKernels.cpp [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+        auto odata = output.packed_accessor64<scalar_t, 4>();
+
+        // Get scaling factors
+        using accscalar_t = at::acc_type<scalar_t, true>;
+        const accscalar_t rheight = area_pixel_compute_scale<accscalar_t>(
+            input_height, output_height, align_corners, scales_h);
+        const accscalar_t rwidth = area_pixel_compute_scale<accscalar_t>(
+            input_width, output_width, align_corners, scales_w);
+
+        upsample_bicubic2d_out_frame<scalar_t, accscalar_t>(
```
Comment (fengyuan14) [inline src/ATen/native/xpu/sycl/UpSampleBicubic2dKernels.cpp:174]: CUDA upsample_bicubic2d_out_frame means a CUDA kernel, not a launcher. Maybe we can name it upsample_bicubic2d_out_template or launch_upsample_bicubic2d_kernel.

File: test/xpu/extended/run_test_with_skip.py [hunk 14]
```diff
@@ -93,6 +91,16 @@
...
+
+    # Not implemented operators
+    "nn_functional_interpolate_linear",
+    "nn_functional_interpolate_bilinear",
+    "nn_functional_interpolate_trilinear",
+
+    # XPU kernel uses accumulate datatype with high accuracy
+    # Not in CUDA test scope too
+    "test_compare_cpu_nn_functional_interpolate_bicubic_xpu_bfloat16",
+    "test_compare_cpu_nn_functional_interpolate_bicubic_xpu_float16",
```
Comment (fengyuan14) [inline test/xpu/extended/run_test_with_skip.py:103]: Did you try if we don't use acc type, we would be same as CPU?

File: test/xpu/extended/run_test_with_skip.py [hunk 15]
```diff
@@ -93,6 +91,16 @@
     # CPU result is not golden reference
     "test_compare_cpu_nn_functional_group_norm_xpu_bfloat16",
     "test_compare_cpu_nn_functional_group_norm_xpu_float16",
+
+    # Not implemented operators
```
Comment (fengyuan14) [inline test/xpu/extended/run_test_with_skip.py:95]: Elaborate the comment.

File: test/xpu/xpu_test_utils.py [hunk 16]
```diff
@@ -118,6 +118,7 @@
     "cdist",
     "nn.functional.group_norm",
     "bincount",
+    'nn.functional.interpolate',
```
Comment (fengyuan14) [inline test/xpu/xpu_test_utils.py:121]: We have got it in line107?

File: yaml/xpu_functions.yaml [hunk 17]
```diff
@@ -383,6 +383,7 @@ supported:
   - _cdist_forward
   - _pin_memory
   - is_pinned
+  - upsample_bicubic2d.out
```
Comment (fengyuan14) [inline yaml/xpu_functions.yaml:386]: You d better putting it near upsample  related ops. Like in line 377.

File: test/xpu/extended/run_test_with_skip.py (hunk 14 above)
Comment (majing921201) [reply test/xpu/extended/run_test_with_skip.py:103]: > Did you try if we don't use acc type, we would be same as CPU?

no, and my previous conclusion is wrong
I think accuracy lost in half-precision is due to Bicubic interpolate algorithm is complex, which includes huge calculation steps
here is the link for algorithm formula 
https://en.wikipedia.org/wiki/Bicubic_interpolation#Bicubic_convolution_algorithm
bicubic need to do twice below computation
image
I suspect accuracy may reduce during this process.

**Review by fengyuan14 — APPROVED** [review]
